src/librustc/infer/error_reporting/named_anon_conflict.rs

```diff
@@ -0,0 +1,206 @@
+                if late_bound_regions.iter().any(|r| *r == br) {
+                    return false;
+                } else {
+                }
+            }
+            _ => {}
+        }
+
+        // Here we check for the case where anonymous region
+        // corresponds to self and if yes, we display E0312.
+        // FIXME(#42700) - Need to format self properly to
+        // enable E0621 for it.
+        if is_first &&
+           self.tcx
+               .opt_associated_item(scope_def_id)
+               .map(|i| i.method_has_self_argument)
+               .unwrap_or(false) {
+            return false;
+        }
+
+        if let Some(simple_name) = arg.pat.simple_name() {
+            struct_span_err!(self.tcx.sess,
+                             span,
+                             E0621,
+                             "explicit lifetime required in the type of `{}`",
+                             simple_name)
+                    .span_label(arg.pat.span,
+                                format!("consider changing the type of `{}` to `{}`",
+                                        simple_name,
+                                        new_ty))
+                    .span_label(span, format!("lifetime `{}` required", named))
+                    .emit();
+
+        } else {
+            struct_span_err!(self.tcx.sess,
+                             span,
+                             E0621,
+                             "explicit lifetime required in parameter type")
+                    .span_label(arg.pat.span,
+                                format!("consider changing type to `{}`", new_ty))
+                    .span_label(span, format!("lifetime `{}` required", named))
+                    .emit();
+        }
+        return true;
+
+    }
+
+    // This method returns whether the given Region is Anonymous
+    // and returns the DefId corresponding to the region.
+    pub fn is_suitable_anonymous_region(&self, region: Region<'tcx>) -> Option<DefId> {
+
+        match *region {
+            ty::ReFree(ref free_region) => {
+                match free_region.bound_region {
+                    ty::BrAnon(..) => {
+                        let anonymous_region_binding_scope = free_region.scope;
+                        let node_id = self.tcx
+                            .hir
+                            .as_local_node_id(anonymous_region_binding_scope)
+                            .unwrap();
+                        match self.tcx.hir.find(node_id) {
+                            Some(hir_map::NodeItem(..)) |
+                            Some(hir_map::NodeTraitItem(..)) => {
+                                // proceed ahead //
+                            }
+                            Some(hir_map::NodeImplItem(..)) => {
+                                if self.tcx.impl_trait_ref(self.tcx.
+associated_item(anonymous_region_binding_scope).container.id()).is_some() {
```

Comment:
Nit: this indentation is a bit wacky, is this what rustfmt proceduces? I would extract a temporary:

```rust
let container_id = self.tcx.associated_item(anonymous_region_binding_scope).container.id();
if self.tcx.impl_trait_ref(container_id).is_some() {
    ...
}
```